metaflow/tutorials/09-hellokfp/kfp_flow.py

```diff
@@ -0,0 +1,56 @@
+from kfp.components import func_to_container_op
+
+from metaflow import FlowSpec, step, kfp
+from collections import namedtuple
+
+
+@func_to_container_op
+def my_step_op_func(x: str, y: str) -> namedtuple("foo", "r1 r2"):
+    """
+    returns r1_value and r2_value appended to x and y respectively
+    """
+    from collections import namedtuple
+
+    print(f"x input parameter: {x}")
+    print(f"y input parameter: {y}")
+
+    assert x == "x_value"
+    assert y == "y_value"
+
+    return namedtuple("ret", "r1 r2")(*[x + " r1_value", y + " r2_value"])
+
+
+class KfpFlow(FlowSpec):
+    """
+    A Flow that decorates a Metaflow Step with a KFP component
+    """
+
+    @kfp(
+        func=my_step_op_func,
+        kfp_component_inputs=["x", "y"],
+        kfp_component_outputs=["r1", "r2"],
+    )
+    @step
+    def start(self):
+        """
+        kfp.kfp_component_inputs Flow state ["x", "y"] is passed to the KFP component as arguments
+        """
+        self.x = "x_value"
+        self.y = "y_value"
+
+        self.next(self.end)
```

Comment:
The KFP step_op `func` actually runs after the MF step.

Is it more intuitive and Python idiomatic when the KFP step_op `func` runs before the MF step?  Like:
```python
    @step
    def start(self):
        """
        kfp.kfp_component_inputs Flow state ["x", "y"] is passed to the KFP component as arguments
        """
        self.x = "x_value"
        self.y = "y_value"

        self.next(self.end)

    @kfp(
        before_step_op_func=my_step_op_func,  # should we rename func to `before_step_op_func` to be explicit?
        kfp_component_inputs=["x", "y"],
        kfp_component_outputs=["r1", "r2"],
    )
    @step
    def end(self):
        """
        kfp.kfp_component_outputs ["r1", "r2'] are now available as Metaflow Flow state
        """
        print("r1", self.r1)
        print("r2", self.r2)

        assert self.r1 == "x_value r1_value"
        assert self.r2 == "y_value r2_value"

Reply:
or this?
```python
    @step
    def start(self):
        """
        kfp.kfp_component_inputs Flow state ["x", "y"] is passed to the KFP component as arguments
        """
        self.x = "x_value"
        self.y = "y_value"

        self.next_kfp_component(
          container_op_func=my_step_op_func,  # should we rename func to `before_step_op_func` to be explicit?
          kfp_component_inputs=["x", "y"],
          kfp_component_outputs=["r1", "r2"],
        )
        self.next(self.end)

    @step
    def end(self):
        """
        kfp.kfp_component_outputs ["r1", "r2'] are now available as Metaflow Flow state
        """
        print("r1", self.r1)
        print("r2", self.r2)
```

or this?
```python
    @step
    def start(self):
        """
        kfp.kfp_component_inputs Flow state ["x", "y"] is passed to the KFP component as arguments
        """
        self.x = "x_value"
        self.y = "y_value"

        self.next(
          container_op_func=my_step_op_func,  # should we rename func to `before_step_op_func` to be explicit?
          kfp_component_inputs=["x", "y"],
          kfp_component_outputs=["r1", "r2"],
          next=self.end  # what could it be named?
        )

    @step
    def end(self):
        """
        kfp.kfp_component_outputs ["r1", "r2'] are now available as Metaflow Flow state
        """
        print("r1", self.r1)
        print("r2", self.r2)
```

Reply:
I think option 1 is most intuitive and Python idiomatic.
`before_step_op_func` could be confused to mean that Metaflow step `end` comes "before" `my_step_op_func`. Perhaps a better name would be `previous_step_op_func` or `relied_upon_step_op_func`. This ensures all the information in the @kfp decorator pertains to the current (`end`) Metaflow step.
If you choose option 3, perhaps the variable `next` within `self.next` can be renamed to `next_metaflow_step`.

Reply:
updated to:
```python
    @kfp(
        preceding_component=load_component_from_url(URL),
        preceding_component_inputs="op1 op2",
        preceding_component_outputs="result",
    )
    @step
    def end(self):
        pass
```

Reply:
Is it possible to have a list when we have multiple inputs or outputs: `preceding_component_inputs="op1 op2"` -> `preceding_component_inputs=["op1", "op2"]? This is what you have previously and it's easier to read and understand.